TestReconnectFromRead test failed (once)

It only happened once locally, and I haven't seen it happen during travis (perhaps it's a new issue), but `github.com/zero-os/0-Disk/tlog/tlogclient`'s test `TestReconnectFromRead ` failed for me once today, during a `make testgo` run:

```
t=2017-06-08T11:57:12-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65235" module=global caller=redis.go:47
t=2017-06-08T11:57:12-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65236" module=global caller=redis.go:47
t=2017-06-08T11:57:12-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65237" module=global caller=redis.go:47
t=2017-06-08T11:57:12-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65238" module=global caller=redis.go:47
t=2017-06-08T11:57:12-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65239" module=global caller=redis.go:47
t=2017-06-08T11:57:12-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65240" module=global caller=redis.go:47
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65242 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65249 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="failed to send resp to :12345, err:writev tcp 127.0.0.1:65241->127.0.0.1:65249: use of closed network connection" module=global caller=server.go:297
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65250 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="failed to send resp to :12345, err:writev tcp 127.0.0.1:65241->127.0.0.1:65250: use of closed network connection" module=global caller=server.go:297
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65251 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65252 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="failed to send resp to :12345, err:writev tcp 127.0.0.1:65241->127.0.0.1:65252: use of closed network connection" module=global caller=server.go:297
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65253 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=eror msg="connection from 127.0.0.1:65254 dropped with an error: read tcp 127.0.0.1:65241->127.0.0.1:65254: read: connection reset by peer" module=global caller=server.go:107
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65255 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="failed to send resp to :12345, err:writev tcp 127.0.0.1:65241->127.0.0.1:65255: use of closed network connection" module=global caller=server.go:297
t=2017-06-08T11:57:12-0500 lvl=eror msg="connection from 127.0.0.1:65256 dropped with an error: read tcp 127.0.0.1:65241->127.0.0.1:65256: read: connection reset by peer" module=global caller=server.go:107
t=2017-06-08T11:57:12-0500 lvl=info msg="failed to send resp to :12345, err:writev tcp 127.0.0.1:65241->127.0.0.1:65256: use of closed network connection" module=global caller=server.go:297
t=2017-06-08T11:57:12-0500 lvl=info msg="connection from 127.0.0.1:65257 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:12-0500 lvl=info msg="failed to send resp to :12345, err:writev tcp 127.0.0.1:65241->127.0.0.1:65257: use of closed network connection" module=global caller=server.go:297
t=2017-06-08T11:57:17-0500 lvl=info msg="connection from 127.0.0.1:65258 dropped" module=global caller=server.go:105
t=2017-06-08T11:57:17-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65263" module=global caller=redis.go:47
t=2017-06-08T11:57:17-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65264" module=global caller=redis.go:47
t=2017-06-08T11:57:17-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65265" module=global caller=redis.go:47
t=2017-06-08T11:57:17-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65266" module=global caller=redis.go:47
t=2017-06-08T11:57:17-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65267" module=global caller=redis.go:47
t=2017-06-08T11:57:17-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:65268" module=global caller=redis.go:47
t=2017-06-08T11:57:17-0500 lvl=info msg="tlogclient : reconnect from read because of : read tcp 127.0.0.1:65270->127.0.0.1:65269: use of closed network connection" module=global caller=client.go:198
t=2017-06-08T11:57:17-0500 lvl=info msg="connection from 127.0.0.1:65270 dropped" module=global caller=server.go:105
--- FAIL: TestReconnectFromRead (9.02s)
	reconnect_test.go:86: TestReconnectFromRead failed : too long
t=2017-06-08T11:57:26-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:49205" module=global caller=redis.go:47
t=2017-06-08T11:57:26-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:49206" module=global caller=redis.go:47
t=2017-06-08T11:57:26-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:49207" module=global caller=redis.go:47
t=2017-06-08T11:57:26-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:49208" module=global caller=redis.go:47
t=2017-06-08T11:57:26-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:49209" module=global caller=redis.go:47
t=2017-06-08T11:57:26-0500 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:49210" module=global caller=redis.go:47
t=2017-06-08T11:57:31-0500 lvl=info msg="tlogclient : reconnect from read because of : EOF" module=global caller=client.go:198
FAIL
FAIL	github.com/zero-os/0-Disk/tlog/tlogclient	19.096s
```

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -58,6 +58,7 @@ type Client struct {
 	cancelFunc context.CancelFunc
 
 	lastConnected time.Time
+	stopped       bool
 }
 
 // New creates a new tlog client for a vdisk with 'addr' is the tlogserver address.
@@ -85,6 +86,9 @@ func New(addr, vdiskID string, firstSequence uint64, resetFirstSeq bool) (*Clien
 // reconnect to server
 // it must be called under wLock
 func (c *Client) reconnect(closedTime time.Time) (err error) {
+	if c.stopped {
+		return nil
+	}
 	if c.lastConnected.After(closedTime) { // another goroutine made the connection work again
 		return
 	}
@@ -390,6 +394,7 @@ func (c *Client) sendReconnect(sender func() (interface{}, error)) (interface{},
 // Close the open connection, making this client invalid.
 // It is user responsibility to call this function.
 func (c *Client) Close() error {
+	c.stopped = true
 	if c.conn != nil {
 		c.conn.CloseRead() // interrupt the receiver
 	}

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -3,6 +3,7 @@ package tlogclient
 import (
 	"bufio"
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"net"
@@ -21,6 +22,10 @@ const (
 	sendSleepTime    = 500 * time.Millisecond // sleep duration before retrying the Send
 	resendTimeoutDur = 2 * time.Second        // duration to wait before re-send the tlog.
 	readTimeout      = 2 * time.Second
+)
+
+var (
+	errMaxSendRetry = errors.New("max send retry reached")
 )
 
 // Response defines a response from tlog server
@@ -388,7 +393,7 @@ func (c *Client) sendReconnect(sender func() (interface{}, error)) (interface{},
 			okToSend = true
 		}
 	}
-	return nil, err
+	return nil, errMaxSendRetry
 }
 
 // Close the open connection, making this client invalid.

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -97,6 +97,10 @@ func (c *Client) reconnect(closedTime time.Time) (err error) {
 
 // connect to server
 func (c *Client) connect(firstSequence uint64, resetFirstSeq bool) (err error) {
+	if c.conn != nil {
+		c.conn.CloseRead() // interrupt the receiver
+	}
+
 	c.rLock.Lock()
 	defer c.rLock.Unlock()
 
@@ -234,7 +238,7 @@ func (c *Client) recvOne() (*Response, error) {
 
 	// set read deadline, so we don't have deadlock
 	// for rLock
-	c.conn.SetReadDeadline(time.Now().Add(readTimeout))
+	//c.conn.SetReadDeadline(time.Now().Add(readTimeout))
 
 	// decode capnp and build response
 	tr, err := c.decodeBlockResponse(c.rd)
@@ -386,13 +390,20 @@ func (c *Client) sendReconnect(sender func() (interface{}, error)) (interface{},
 // Close the open connection, making this client invalid.
 // It is user responsibility to call this function.
 func (c *Client) Close() error {
+	if c.conn != nil {
+		c.conn.CloseRead() // interrupt the receiver
+	}
+
 	c.cancelFunc()
 
 	c.wLock.Lock()
 	defer c.wLock.Unlock()
 
 	c.rLock.Lock()
-	c.rLock.Unlock()
+	defer c.rLock.Unlock()
 
-	return c.conn.Close()
+	if c != nil {
+		return c.conn.Close()
+	}
+	return nil
 }

